File: frontend/src/utils/achievements.ts
export type Achievement = {
  id: string;
  label: string;
  icon: string;
  description: string;
  bonus: number;
  unlocked: boolean;
};

type AchievementInput = {
  gamesPlayed: number;
  totalScore: number;
  bestRoundScore: number;
};

const DEFINITIONS = [
  {
    id: "first-game",
    label: "Birinchi qadam",
    icon: "\u{1F3AF}",
    description: "1 ta o'yin o'ynang",
    bonus: 5,
    isUnlocked: (input: AchievementInput) => input.gamesPlayed >= 1
  },
  {
    id: "ten-games",
    label: "Tajribali",
    icon: "\u{1F3AE}",
    description: "10 ta o'yin o'ynang",
    bonus: 15,
    isUnlocked: (input: AchievementInput) => input.gamesPlayed >= 10
  },
  {
    id: "score-100",
    label: "Zukko",
    icon: "\u{1F9E0}",
    description: "100 ball to'plang",
    bonus: 20,
    isUnlocked: (input: AchievementInput) => input.totalScore >= 100
  },
  {
    id: "score-500",
    label: "Daho",
    icon: "\u{1F31F}",
    description: "500 ball to'plang",
    bonus: 50,
    isUnlocked: (input: AchievementInput) => input.totalScore >= 500
  },
  {
    id: "best-round",
    label: "Rekordchi",
    icon: "\u{1F3C6}",
    description: "Bir raundda 20+ ball",
    bonus: 25,
    isUnlocked: (input: AchievementInput) => input.bestRoundScore >= 20
  }
];

export function computeAchievements(input: AchievementInput): Achievement[] {
  return DEFINITIONS.map(({ isUnlocked, ...rest }) => ({
    ...rest,
    unlocked: isUnlocked(input)
  }));
}

File: backend-django/apps/users/achievements.py
"""Yutuqlar va ularning bonus ballari — server-side ro'yxat.

Frontend dagi `utils/achievements.ts` bilan bir xil sinxron bo'lishi kerak.
Server xavfsizroq: foydalanuvchi localStorage'ni o'zgartirib bonusni
takror olishga uddasi yo'q — bonus faqat bir marta beriladi, chunki
`unlocked_achievements` ro'yxati bazaga yoziladi.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable


@dataclass(frozen=True)
class AchievementDef:
    id: str
    label: str
    bonus: int
    is_unlocked: Callable[[dict], bool]


ACHIEVEMENTS: list[AchievementDef] = [
    AchievementDef(
        id="first-game",
        label="Birinchi qadam",
        bonus=5,
        is_unlocked=lambda stats: stats.get("gamesPlayed", 0) >= 1,
    ),
    AchievementDef(
        id="ten-games",
        label="Tajribali",
        bonus=15,
        is_unlocked=lambda stats: stats.get("gamesPlayed", 0) >= 10,
    ),
    AchievementDef(
        id="score-100",
        label="Zukko",
        bonus=20,
        is_unlocked=lambda stats: stats.get("totalScore", 0) >= 100,
    ),
    AchievementDef(
        id="score-500",
        label="Daho",
        bonus=50,
        is_unlocked=lambda stats: stats.get("totalScore", 0) >= 500,
    ),
    AchievementDef(
        id="best-round",
        label="Rekordchi",
        bonus=25,
        is_unlocked=lambda stats: stats.get("bestRoundScore", 0) >= 20,
    ),
]


def find_newly_unlocked(stats: dict, already_unlocked: list[str]) -> list[AchievementDef]:
    seen = set(already_unlocked)
    return [a for a in ACHIEVEMENTS if a.id not in seen and a.is_unlocked(stats)]
